packages/eslint-plugin/src/rules/no-confusing-void-expression.ts

```diff
@@ -376,5 +399,33 @@ export default createRule<Options, MessageId>({
       const type = getConstrainedTypeAtLocation(services, targetNode);
       return tsutils.isTypeFlagSet(type, ts.TypeFlags.VoidLike);
     }
+
+    function targetNodeFunctionAncestorNodeHasVoidReturnType(
+      node: TSESTree.Node,
+    ): boolean {
+      if (!node.parent) {
+        return false;
+      }
+
+      if (
+        AST_NODE_TYPES.FunctionDeclaration === node.parent.type ||
+        AST_NODE_TYPES.ArrowFunctionExpression === node.parent.type ||
+        AST_NODE_TYPES.FunctionExpression === node.parent.type
+      ) {
+        const services = getParserServices(context);
+        if (
+          node.parent.returnType &&
+          !tsutils.isTypeFlagSet(
+            getConstrainedTypeAtLocation(services, node.parent.returnType),
+            ts.TypeFlags.VoidLike,
+          )
+        ) {
+          return true;
+        }
+        return false;
```

Comment:
```suggestion
        return (
          node.parent.returnType &&
          !tsutils.isTypeFlagSet(
            getConstrainedTypeAtLocation(services, node.parent.returnType),
            ts.TypeFlags.VoidLike,
          )
        )
```

[Refactor] This code may be converted into a single return statement. What do you think?